lib/puppet/util/character_encoding.rb

```diff
@@ -13,12 +12,16 @@ class << self
     # @raise [Puppet::Error] a failure to legitimately set external encoding or
     #   transcode string
     def convert_to_utf_8(string)
-      return string if string.encoding == Encoding::UTF_8
-
-      value_to_encode = string.dup
+      return string if (string.encoding == Encoding::UTF_8 && string.valid_encoding?)
 
       begin
-        if valid_utf_8?(value_to_encode)
+        if already_invalid_utf_8?(string)
+          # If a string is currently believed to be UTF-8, but is also not
+          # valid_encoding?, we have no recourse but to fail because we have no
+          # idea what encoding this string originally came from where it *was*
+          # valid - all we know is it's not currently valid UTF-8.
+          raise EncodingError
```

Comment:
I think this method could be simplified so we don't call `valid_encoding?` more than once (even though it looks like ruby will memoize the result, until the string is next mutated). The reason I mention that is `valid_encoding?` will at worst case scan the string multiple times (once to see if there are any non-US ASCII bytes, and then again looking for multi-byte sequences). So maybe something like?

```ruby
valid = string.valid_encoding?

if string.encoding == Encoding::UTF_8
  if valid
    string
  else
    raise EncodingError
  end
elsif valid_utf_8?(string)
  string.force_encoding(Encoding::UTF_8)
else
  string.encode!(Encoding::UTF_8)
end
```

That complexity is because we're trying to figure out if the string is already valid UTF-8 or could be force encoded. It would take a lot of string encodings for that to matter I think. For example, if I run a benchmark 10M times, it takes 4.2 seconds to create the various string objects. If I also try to UTF-8 encode a string already in UTF-8 encoding, it takes 7.57 seconds. So it's not a noop, but it takes about as long to "noop encode" the string as it does to create the string itself, probably due to making a copy of the string. But if I UTF-8 encode a string that's in ISO-8859-1 (even though the byte sequences are exactly the same), it takes over 10x times as long. So ruby is doing a decent job at short-circuiting the encoding logic when source and destination encodings are the same.

```
$ ruby enc.rb
            baseline  4.220000   0.070000   4.290000 (  4.279521)
      same encodings  7.570000   0.010000   7.580000 (  7.591500)
      diff encodings 81.410000   0.750000  82.160000 ( 82.362164)
```

If we wanted to simplify the logic, we could do something like:

```ruby
begin
  string.encode!(Encoding::UTF_8)
rescue EncodingError => e
  original_encoding = string.encoding
  string.force_encoding(Encoding::UTF_8)
  if string.valid_encoding?
    string
  else
    string.force_encoding(original_encoding)
    raise e
  end
end
```